### PR TITLE
test(dot/parachain/backing): we can second a valid candidate when the previous candidate has been found invalid

### LINKED ISSUE #4007: test(dot/parachain/backing): if we are asked to second an invalid candidate we can still second a valid one afterwards.
When we get a Second message in the candidate backing subsystem.
- it should validate the candidate
    - if it's an Invalid candidate, it should not be seconded. Instead, this subsystem should report to the collator protocol about invalidity.
    
- afterwards, if we get another message to Second the candidate, If it's a valid candidate
    - If backing is allowed, the subsystem should
        - Sign the statement
        - Import the statement into the statement table
        - and finally, issue a message to distribute the statement

---
- **candidate**: It is a submitted parachain block to the relay chain validators. A parachain block stops being referred to as a candidate as soon it has been finalized
- **Statement** could be either `Valid` or `Seconded`. (Seconded for this test)
    - **Seconded**: The block is put forward by the para-validator to other para-validators (Implicit validity vote)
    - **Valid**: statement that a validator has deemed a candidate valid

### PR DESCRIPTION
**Pull request by axaysagathiya**
## Changes
- The new integration test is in the `dot/parachain/backing/integration_test.go` file
- Exported some fields of the `CandidateBacking` struct because I wanted to mock them
- modified the logic of `processMessages()` method of `MockableOverseer` struct

## We have tested
When we get a message to second an Invalid candidate,
- don't second the candidate.
- report to collator protocol about an invalid candidate

After this, if we get another message to second the valid candidate
- second the candidate
- store available data
- distribute the statement to peers
- inform collator protocol about the seconded candidate 

## Tests

```sh
go test -tags integration github.com/ChainSafe/gossamer
```

## Issues
Closes #4007

### CODE REVIEW

File: dot/parachain/overseer/mockable_overseer.go [hunk 1]
```diff
@@ -50,14 +48,15 @@ func (m *MockableOverseer) Start() error {
 		sub.Run(m.ctx, overseerToSubSystem, m.SubsystemsToOverseer)
 	}(m.subSystem, m.overseerToSubsystem)
 
-	go m.processMessages()
+	// go m.processMessages()
```
Comment (axaysagathiya) [inline dot/parachain/overseer/mockable_overseer.go]: @kishansagathiya 

I couldn't figure out how to use this properly.
So now, to mock the actions of the overseer(and other subsystems), I have goroutines in the test.

File: dot/parachain/availability-store/messages.go [hunk 2]
```diff
@@ -92,8 +92,8 @@ type StoreAvailableData struct {
 
 // AvailableData is the data we keep available for each candidate included in the relay chain
 type AvailableData struct {
-	PoV            parachaintypes.PoV
-	ValidationData parachaintypes.PersistedValidationData
+	PoV            parachaintypes.PoV                     `scale:"1"`
+	ValidationData parachaintypes.PersistedValidationData `scale:"2"`
 }
```
Comment (kishansagathiya) [inline dot/parachain/availability-store/messages.go:97]: I like it that we explicitly add the scale tag.

But keep in mind that these scale tags are not compulsory. If you don't add them, all the properties have an implied scale tag based on their order.

File: dot/parachain/backing/active_leaves_update.go [hunk 3]
```diff
@@ -218,7 +218,7 @@ func (cb *CandidateBacking) cleanUpPerRelayParentByLeafAncestry() {
 		remaining[hash] = true
 	}
 
-	allowedRelayParents := cb.implicitView.allAllowedRelayParents()
+	allowedRelayParents := cb.ImplicitView.AllAllowedRelayParents()
```
Comment (kishansagathiya) [inline dot/parachain/backing/active_leaves_update.go:221]: I tried keeping `allAllowedRelayParents` unexported and everything still works. You probably don't need to export it.

File: dot/parachain/backing/backing_implicit_view.go [hunk 4]
```diff
@@ -13,8 +13,8 @@ import (
...
 // parachains at those leaves
 type ImplicitView interface {
-	knownAllowedRelayParentsUnder(blockHash common.Hash, paraID *parachaintypes.ParaID) []common.Hash
-	activeLeaf(leafHash common.Hash) ([]parachaintypes.ParaID, error)
-	deactivateLeaf(leafHash common.Hash) []common.Hash
-	allAllowedRelayParents() []common.Hash
+	KnownAllowedRelayParentsUnder(blockHash common.Hash, paraID *parachaintypes.ParaID) []common.Hash
+	ActiveLeaf(leafHash common.Hash) ([]parachaintypes.ParaID, error)
+	DeactivateLeaf(leafHash common.Hash) []common.Hash
+	AllAllowedRelayParents() []common.Hash
```
Comment (kishansagathiya) [inline dot/parachain/backing/backing_implicit_view.go]: I don't see the point of this change. I unexported all these methods and `make gossamer` and `go test ./dot/parachain/backing/...` still works.

File: dot/parachain/overseer/mockable_overseer.go (hunk 1 above)
Comment (kishansagathiya) [reply dot/parachain/overseer/mockable_overseer.go]: This is how I have used it to write a test in collator protocol.

```
		        mockOverseer := overseer.NewMockableOverseer(t)
			overseerToSubsystem := mockOverseer.RegisterSubsystem(&cpvs)
			cpvs.OverseerToSubSystem = overseerToSubsystem
			cpvs.SubSystemToOverseer = mockOverseer.GetSubsystemToOverseerChannel()

			mockOverseer.Start()
			defer mockOverseer.Stop()

			mockOverseer.ExpectMessageWithAction(c.expectedMessage, func(msg any) {
				// do nothing
			})
			msg := collatorprotocolmessages.NewCollationProtocol()
			vdtChild := collatorprotocolmessages.NewCollatorProtocolMessage()
```

Let me look at your code and see if I can use mockable overseer there.

File: go.mod
Comment (kishansagathiya) [inline go.mod:1]: is `go mod tidy` causing this?

File: dot/parachain/overseer/mockable_overseer.go (hunk 1 above)
Comment (axaysagathiya) [reply dot/parachain/overseer/mockable_overseer.go]: I had to change the logic of the `processMessages()` method of mockable overseer.
I have pushed the updated code.

File: dot/parachain/backing/active_leaves_update.go (hunk 3 above)
Comment (axaysagathiya) [reply dot/parachain/backing/active_leaves_update.go:221]: I am mocking this method in a different package. So I need this exported.

File: dot/parachain/backing/backing_implicit_view.go (hunk 4 above)
Comment (axaysagathiya) [reply dot/parachain/backing/backing_implicit_view.go:19]: I am mocking some of these methods in this PR, So I need them exported.

File: go.mod
Comment (axaysagathiya) [reply go.mod:1]: Yes

File: dot/parachain/backing/backing_implicit_view.go (hunk 4 above)
Comment (axaysagathiya) [reply dot/parachain/backing/backing_implicit_view.go]: I have mocked `AllAllowedRelayParents()` in this PR. So, we need this method exported as I am writing this test in `backing_test` package

unexported other methods.

File: dot/parachain/overseer/mockable_overseer.go [hunk 5]
```diff
@@ -19,18 +19,19 @@ type MockableOverseer struct {
 	overseerToSubsystem  chan any
 	subSystem            parachaintypes.Subsystem
 
-	expectedMessagesWithAction map[any]func(msg any)
+	// expected actions for overseer messages we receive from the subsystem.
+	// need to return false if the message is unexpected
+	actions []func(msg any) bool
```
Comment (kishansagathiya) [inline dot/parachain/overseer/mockable_overseer.go]: keep a similar name like `actionsForExpectedMessages`

File: dot/parachain/overseer/mockable_overseer.go [hunk 6]
```diff
@@ -19,18 +19,19 @@ type MockableOverseer struct {
...
 	return &MockableOverseer{
-		t:                          t,
-		ctx:                        ctx,
-		cancel:                     cancel,
-		SubsystemsToOverseer:       make(chan any),
-		expectedMessagesWithAction: make(map[any]func(msg any)),
+		t:                    t,
+		ctx:                  ctx,
+		cancel:               cancel,
+		SubsystemsToOverseer: make(chan any),
```
Comment (kishansagathiya) [inline dot/parachain/overseer/mockable_overseer.go]: initialize the actions slice as well

File: dot/parachain/overseer/mockable_overseer.go [hunk 7]
```diff
@@ -56,30 +57,40 @@ func (m *MockableOverseer) Start() error {
...
 }
 
+// ReceiveMessage method is to receive overseer messages in a subsystem which we are testing
 func (m *MockableOverseer) ReceiveMessage(msg any) {
 	m.overseerToSubsystem <- msg
 }
 
-func (m *MockableOverseer) ExpectMessageWithAction(msg any, fn func(msg any)) {
-	m.expectedMessagesWithAction[msg] = fn
+// ExpectActions method is to set expected actions for overseer messages we receive from the subsystem.
```
Comment (kishansagathiya) [inline dot/parachain/overseer/mockable_overseer.go]:
```suggestion
// ExpectActions is to set expected actions for overseer messages we receive from the subsystem.
```

File: dot/parachain/overseer/mockable_overseer.go (hunk 5 above)
Comment (kishansagathiya) [inline dot/parachain/overseer/mockable_overseer.go]: add a note here that

actions must be in the order that we expect them. At some point in future if we can't be sure of in which order messages will be sent (say multiple parallel tasks running that could emit messages at any time) , we will need to modify this.

File: dot/parachain/overseer/mockable_overseer.go [hunk 8]
```diff
@@ -56,30 +57,40 @@ func (m *MockableOverseer) Start() error {
...
 		select {
 		case msg := <-m.SubsystemsToOverseer:
 			if msg == nil {
 				continue
 			}
-			action, ok := m.expectedMessagesWithAction[msg]
-			if !ok {
-				m.t.Errorf("unexpected message: %v", msg)
-				continue
-			}
```
Comment (kishansagathiya) [inline dot/parachain/overseer/mockable_overseer.go:81]: why not t.Error in case of an unexpected message?

Comment (axaysagathiya) [reply dot/parachain/overseer/mockable_overseer.go:81]: If you look at the updated code, we are doing it; it has not been removed.

**Review by edwardmack — APPROVED** [review]

**Review by kishansagathiya — APPROVED** [review]

**Pull request merged**